Deploy website + mixer + esp on Kubernetes (#698)
datacommonsorg/website: deployment/minikube/README.md

# Bring up the Website App locally in Minikube

> Website can be deployed in Kubenetes cluster on [GKE](https://cloud.google.com/kubernetes-engine) or

Author: Done

[MiniKube](https://minikube.sigs.k8s.io/docs/).

This folder contains the config and script for local development with Minikube.

> ## Prerequisit

Author: Done

> - Install `Docker`, following [Docker Desktop](https://www.docker.com/products/docker-desktop) install guide.

Author: Done

> - Install `Minikube`, following the [Installation](https://minikube.sigs.k8s.io/docs/start/) guide.

Reviewer: Some installation steps have "following the installation" while others don't, removing for consistency.

> - Install `Minikube`, following the [Installation](https://minikube.sigs.k8s.io/docs/start/) guide.

Author: Updated

- Install [`kubectl`](https://kubernetes.io/docs/tasks/tools/install-kubectl/).

- Install [`yq`](https://mikefarah.gitbook.io/yq/).

> Set GCP roject

Reviewer: ```suggestion
Set GCP project
```

```bash
export GCP_PROJECT="<YOUR_PROJECT>"
```

GCP setup

```bash
gcloud config set project $GCP_PROJECT
gcloud components update
gcloud auth login
```

- If not done yet, run gcloud application credential and copy the auth key:

```bash
gcloud auth application-default login
gcloud iam service-accounts keys create website-robot-key.json \
      --iam-account mixer-robot@$GCP_PROJECT.iam.gserviceaccount.com
# Use the same robot account for website and mixer
cp website-robot-key.json mixer-robot-key.json
```

## Update Mixer Submodule

This repo contains DataCommons [mixer repo](https://github.com/datacommonsorg/mixer) as a submodule.

Need to track the upstream change:

```bash
git submodule foreach git pull origin master
```

## Run Minikube cluster

From a terminal, start a cluster:

```bash
minikube start --memory=4g
```

> Use local docker image with Minikube.

Reviewer: ```suggestion
Use local Docker image with Minikube.
```

> **NOTE** Need to run "docker build" after this command to make them usable in deployment.yaml

```bash
eval $(minikube docker-env)
```

Start Minikube dashboard

```bash
minikube dashboard
```

Create a new namespace "website"

```bash
kubectl create namespace website
```

Mount the GCP credential

```bash
kubectl create secret generic website-robot-key --from-file=website-robot-key.json --namespace=website
kubectl create secret generic mixer-robot-key --from-file=mixer-robot-key.json --namespace=website
```

## Generate YAML files

```bash
./generate_yaml.sh
```

## Deploy ESP configuration [One time]

```bash
gsutil cp gs://artifacts.datcom-ci.appspot.com/mixer-grpc/mixer-grpc.latest.pb .
gcloud endpoints services deploy mixer-grpc.latest.pb endpoints.yaml
. env.sh
gcloud services enable $SERVICE_NAME
```

> ## Build Docker Image [Run after code change]

Reviewer: ```suggestion
## Build Docker image [Run after code change]
```

> The build would take a few mintues the first time. Subsquent build should only take a few seconds with docker caching.

Reviewer: ```suggestion
The build should take a few minutes the first time. Subsequent builds should only take a few seconds with docker caching.
```

# Bring up the Website App locally in Minikube

Website can be deployed in Kubernetes cluster on [GKE](https://cloud.google.com/kubernetes-engine) or
[MiniKube](https://minikube.sigs.k8s.io/docs/).

This folder contains the config and script for local development with Minikube.

## Prerequisites

- Install [`Docker`](https://www.docker.com/products/docker-desktop).

- Install [`Minikube`](https://minikube.sigs.k8s.io/docs/start/).

- Install [`kubectl`](https://kubernetes.io/docs/tasks/tools/install-kubectl/).

- Install [`yq`](https://mikefarah.gitbook.io/yq/).

## One time setup

```bash
./onetime_setup.sh
```

## Deploy endpoints configuration

> **Note** Need to run this regularly to keep track of the mixer service update.

Reviewer: ```suggestion
**Note** Need to run this regularly to keep track of the mixer service updates.
```

Ideally, mixer-grpc.pb could be generated locally with protoc, or check the mixer.pto change and only deploy esp service when there is proto change.

```bash
./run_endponit_service.sh
```

## Run Minikube cluster

> This will bring up the minikube cluster and let it running.

Reviewer: ```suggestion
This will bring up the minikube cluster and keep it running.
```

> You can let it running in the background or stop it after development to save 4G of memory on you computer.

Reviewer: ```suggestion
You can keep it running in the background or stop it after development to save 4G of memory on you computer.
```
